### PR TITLE
[reconfigurator] ClickHouse XML generation types

### PR DESCRIPTION
**Pull request by karencfv**
## Overview

This commit adds a library to generate ClickHouse replica server and keeper configuration files. A lot of the code in the `clickhouse-admin/types` directory has been copied over from [clickward](https://github.com/oxidecomputer/clickward), but there are several additions and modifications:

- New `new()` and `default()` methods that set default Oxide values.
- File generation is per node, as opposed to all files generated in a single directory like clickward.

## Usage

To generate a replica server configuration file:

```rust
let keepers = vec![
    KeeperNodeConfig::new("ff::01".to_string()),
    KeeperNodeConfig::new("ff::02".to_string()),
    KeeperNodeConfig::new("ff::03".to_string()),
];

let servers = vec![
    ServerNodeConfig::new("ff::08".to_string()),
    ServerNodeConfig::new("ff::09".to_string()),
];

let config = ClickhouseServerConfig::new(
    Utf8PathBuf::from(config_dir.path()),
    ServerId(1),
    Utf8PathBuf::from_str("./").unwrap(),
    Ipv6Addr::from_str("ff::08").unwrap(),
    keepers,
    servers,
);

config.generate_xml_file().unwrap();
```

To generate a keeper configuration file:

```rust
let keepers = vec![
    RaftServerConfig::new(KeeperId(1), "ff::01".to_string()),
    RaftServerConfig::new(KeeperId(2), "ff::02".to_string()),
    RaftServerConfig::new(KeeperId(3), "ff::03".to_string()),
];

let config = ClickhouseKeeperConfig::new(
    Utf8PathBuf::from(config_dir.path()),
    KeeperId(1),
    keepers,
    Utf8PathBuf::from_str("./").unwrap(),
    Ipv6Addr::from_str("ff::08").unwrap(),
);

config.generate_xml_file().unwrap();
```

## Purpose

As part of the work to roll out replicated ClickHouse, we'll need to dynamically generate the node configuration files via the `clickhouse-admin` service. This commit is part of the work necessary to do so.

Related: https://github.com/oxidecomputer/omicron/issues/5999 , https://github.com/oxidecomputer/omicron/issues/3824

### CODE REVIEW

**Review by andrewjstone — APPROVED** [review]
Karen, this looks great! Just some minor comments.

File: clickhouse-admin/types/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    #[test]
+    fn test_generate_replica_config() {
+        let config_dir = Builder::new()
+        .tempdir_in(
+            Utf8PathBuf::try_from(std::env::temp_dir()).unwrap())
+            .expect("Could not create directory for ClickHouse configuration generation test"
+        );
+
+        let keepers = vec![
+            KeeperNodeConfig::new("ff::01".to_string()),
```
[the remark this answers is not in the document]
Comment (karencfv) [reply clickhouse-admin/types/src/lib.rs:206]: Ha! I [ran](https://github.com/oxidecomputer/omicron/pull/4372#issuecomment-1786419465) [into](https://github.com/oxidecomputer/omicron/blob/main/test-utils/src/dev/clickhouse.rs#L173-L177) this annoying ClickHouse [quirk](https://github.com/ClickHouse/ClickHouse/blob/a011990fd75628c63c7995c4f15475f1d4125d10/src/Coordination/KeeperStateManager.cpp#L149) some time ago as well. It appears I forgot about it 😅 .

**Review by karencfv — COMMENTED** [review]
Thanks for the review @andrewjstone! I'll update the PR with your comments

File: clickhouse-admin/types/src/config.rs [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+        Self { servers }
+    }
+    pub fn to_xml(&self) -> String {
+        let mut s = String::new();
+        for server in &self.servers {
+            let RaftServerConfig { id, hostname, port } = server;
+            s.push_str(&format!(
+                "
+            <server>
+                <id>{id}</id>
```
[the remark this answers is not in the document]
Comment (andrewjstone) [reply clickhouse-admin/types/src/config.rs:385]: "I'll show you where the change should go".
... Proceeds to show the wrong location

**Pull request merged**